tests/test-portal.c

```diff
@@ -0,0 +1,496 @@
+
+  g_subprocess_communicate_utf8 (f->portal, NULL, NULL, &stdout_buf,
+                                 &stderr_buf, &error);
+  g_assert_nonnull (stdout_buf);
+  g_assert_nonnull (stderr_buf);
+  g_test_message ("flatpak-portal --help: %s", stdout_buf);
+  g_assert_true (strstr (stdout_buf, "--replace") != NULL);
+
+  g_subprocess_wait_check (f->portal, NULL, &error);
+  g_assert_no_error (error);
+}
+
+static void
+count_successful_exit_cb (PortalFlatpak *proxy,
+                          guint pid,
+                          guint wait_status,
+                          gpointer user_data)
+{
+  gsize *times_exited_p = user_data;
+
+  g_debug ("Process %u exited with wait status %u", pid, wait_status);
+  g_assert_true (WIFEXITED (wait_status));
+  g_assert_cmpuint (WEXITSTATUS (wait_status), ==, 0);
+  (*times_exited_p) += 1;
+}
+
+static void
+test_basic (Fixture *f,
+            gconstpointer context G_GNUC_UNUSED)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GUnixFDList) fds_out = NULL;
+  glnx_autofd int stderr_copy = -1;
+  guint pid;
+  gboolean ok;
+  const char * const argv[] = { "hello", NULL };
+  gsize times_exited = 0;
+  gulong handler_id;
+
+  fixture_start_portal (f);
+
+  /* We can't easily tell whether EXPOSE_PIDS ought to be set or not */
+  g_assert_cmpuint ((portal_flatpak_get_supports (f->proxy) &
+                     (~FLATPAK_SPAWN_SUPPORT_FLAGS_EXPOSE_PIDS)), ==, 0);
+  g_assert_cmpuint (portal_flatpak_get_version (f->proxy), ==, 6);
+
+  handler_id = g_signal_connect (f->proxy, "spawn-exited",
+                                 G_CALLBACK (count_successful_exit_cb),
+                                 &times_exited);
+
+  ok = portal_flatpak_call_spawn_sync (f->proxy,
+                                       "/",           /* cwd */
+                                       argv,          /* argv */
+                                       g_variant_new ("a{uh}", NULL),
+                                       g_variant_new ("a{ss}", NULL),
+                                       FLATPAK_SPAWN_FLAGS_NONE,
+                                       g_variant_new ("a{sv}", NULL),
+                                       NULL,          /* fd list */
+                                       &pid,
+                                       &fds_out,
+                                       NULL,
+                                       &error);
+  g_assert_no_error (error);
+  g_assert_true (ok);
+  g_assert_cmpuint (pid, >, 1);
+
+  while (times_exited == 0)
+    g_main_context_iteration (NULL, TRUE);
+
+  g_signal_handler_disconnect (f->proxy, handler_id);
+
+  if (fds_out != NULL)
+    g_assert_cmpint (g_unix_fd_list_get_length (fds_out), ==, 0);
+
+  g_subprocess_send_signal (f->portal, SIGTERM);
+  g_subprocess_wait (f->portal, NULL, &error);
+  g_assert_no_error (error);
+}
+
+static void
+test_fd_passing (Fixture *f,
+                 gconstpointer context G_GNUC_UNUSED)
+{
+#define SOME_FDS 16
```

Comment:
This is as many as dbus-daemon will allow through by default.

Reply:
Is it safe to rely on this hardcoded constant? Especially with dbus-broker and whatnot existing.

Reply:
`setup_dbus_daemon()` is specifically using a private instance of dbus-daemon, not whatever the OS uses for actual user sessions (which might be dbus-broker). There's unfortunately no API to discover the maximum number of fds allowed in a message, but I suppose we could supply a dbus-daemon configuration file that forces a particular value?

The lower we make this, the harder it'll be to reproduce #4286. We need a value chosen so that the "source" fds attached to the D-Bus message, plus the other fds that are approximately constant for any given versions of Flatpak and GLib (stdin/stdout/stderr, the D-Bus connection, and any other fds we have open for other reasons), will push the next fd to be allocated into somewhere in the "destination" range determined by `gap_size`.

Reply:
Updated version uses `tests/session.conf.in`, which sets the maximum number of fds to more than this.